helper/plugin/grpc_provider.go

```diff
@@ -261,30 +261,31 @@ func (s *GRPCProviderServer) UpgradeResourceState(_ context.Context, req *proto.
 
 	res := s.provider.ResourcesMap[req.TypeName]
 	blockForCore := s.getResourceSchemaBlockForCore(req.TypeName)
-	blockForShimming := s.getResourceSchemaBlockForShimming(req.TypeName)
 
 	version := int(req.Version)
 
 	var jsonMap map[string]interface{}
 	var err error
 
-	// if there's a JSON state, we need to decode it.
-	if len(req.RawState.Json) > 0 {
-		err = json.Unmarshal(req.RawState.Json, &jsonMap)
+	switch {
+	// We first need to upgrade a flatmap state if it exists.
+	// There should never be both a JSON and Flatmap state in the request.
+	case len(req.RawState.Flatmap) > 0:
```

Comment:
I think we had these the other way around originally because an empty `req.RawState.Json` is never valid (it should always be at least `{}`) but yet an empty `req.RawState.Flatmap` _can_ potentially be valid, for an empty schema.

With that said, I suppose in the old SDK a schema is never truly empty because the `id` attribute is always implied and can never be left unset, so perhaps it doesn't matter.

Reply:
Yeah, we shouldn't see a state without an id in the legacy SDK, so I moved this around to make it a little easier to understand. While it shouldn't happen in core, it's not enforced that req.RawState.Flatmap must be nil, and an empty flatmap state it could have overwritten a valid json state.